pkg/common/workbuilder/workbuilder.go

```diff
@@ -0,0 +1,318 @@
+	for _, work := range appliedWorks {
+		rst, err := w.workApplier.Apply(ctx, &work)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			resultWorks = append(resultWorks, *rst)
+		}
+	}
+
+	for _, work := range deletedWorks {
+		if err = w.workApplier.Delete(ctx, work.Namespace, work.Name); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) != 0 {
+		return resultWorks, fmt.Errorf("failed to apply all works. errs: %v", errs)
+	}
+	return resultWorks, nil
+
+}
+
+func (w *WorkBuilder) newManifestWorks(manifests []workapiv1.Manifest, workIndex int) ([]workapiv1.ManifestWork, error) {
+	var manifestWorks []workapiv1.ManifestWork
+	// sort from big to small by size
+	sort.SliceStable(manifests, func(i, j int) bool {
+		return manifests[j].Size() < manifests[i].Size()
+	})
+
+	var totalSize = 0
+	for i := 0; i < len(manifests); i++ {
+		if totalSize == 0 {
+			work := w.generateManifestWork(workIndex)
+			work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, manifests[i])
+			totalSize = manifests[i].Size()
+			manifestWorks = append(manifestWorks, work)
+			continue
+		}
+		if totalSize+manifests[i].Size() < w.manifestLimit {
+			work := &manifestWorks[len(manifestWorks)-1]
+			work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, manifests[i])
+			totalSize = totalSize + manifests[i].Size()
+			continue
+		}
+
+		workIndex = workIndex + 1
+		work := w.generateManifestWork(workIndex)
+		work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, manifests[i])
+		totalSize = manifests[i].Size()
+		manifestWorks = append(manifestWorks, work)
+	}
+
+	return manifestWorks, nil
+}
+
+func (w *WorkBuilder) buildManifestWorks(objects []runtime.Object,
+	existingManifestWorks []workapiv1.ManifestWork) (appliedWorks, deletedWorks []workapiv1.ManifestWork, err error) {
+	if len(existingManifestWorks) == 0 {
+		requiredWorks, err := w.NewBuild(objects)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to new build works. err: %v", err)
+		}
+		return requiredWorks, nil, nil
+	}
+
+	requireMapper, err := generateRequiredManifestMapper(objects)
+	if err != nil {
+		return nil, nil, err
+	}
+	return w.compareManifests(existingManifestWorks, requireMapper)
+}
+
+func generateManifestKey(manifest workapiv1.Manifest) (manifestKey, error) {
+	var object runtime.Object
+	var err error
+	if manifest.Object != nil {
+		object = manifest.Object
+	} else {
+		object, err = runtime.Decode(unstructured.UnstructuredJSONScheme, manifest.Raw)
+		if err != nil {
+			return manifestKey{}, err
+		}
+	}
+
+	key := manifestKey{
+		gvk: object.GetObjectKind().GroupVersionKind(),
+	}
+
+	accessor, err := meta.Accessor(object)
+	if err != nil {
+		return key, err
+	}
+	key.namespace = accessor.GetNamespace()
+	key.name = accessor.GetName()
+	return key, nil
+}
+
+func buildManifest(object runtime.Object) (workapiv1.Manifest, error) {
+	rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
```

Comment:
are you sure it works? what if the object is not unstructure?

Reply:
verified it works.